CI silently swallows errors and tests forget to check return values

Here is a simple example that has blocked my PR #9788 for weeks. 

In this PR, I need to use `wasm32-wasip2` target, which was not added here
https://github.com/bytecodealliance/wasmtime/blob/fc7ef8d62e196df90522b26ed055bc21667cd19a/.github/actions/install-rust/action.yml#L67
which is used by CI.

My PR added L69 to compile an example with `wasm32-wasip2` target like below:

<img width="1199" alt="image" src="https://github.com/user-attachments/assets/f5780e6a-0e2e-4b67-a85c-b576ec431b38" />

This `create_rust_wasm(wasi wasm32-wasip2)` should have failed CI but not, because it didn't checked the return code of the command that was run.

The command should have returned something like 
```shell
cargo build -p example-wasi-wasm --target wasm32-wasip2
   Compiling example-wasi-wasm v0.0.0 (/Users/zhiqiu/offline_code/opensource/wasmtime/examples/wasm)
error[E0463]: can't find crate for `std`
  |
  = note: the `wasm32-wasip2` target may not be installed
  = help: consider downloading the target with `rustup target add wasm32-wasip2`

For more information about this error, try `rustc --explain E0463`.
error: could not compile `example-wasi-wasm` (bin "wasi") due to 1 previous error
```

This error was not checked, so a test later failed with misleading error messages, because this step didn't compiled a required manifest file (`wasi.wasm`).

I don't know if this case is a special case or there are loopholes in CI in which CMake is used but it does not checked a return code of a command that is run.

--- a/examples/CMakeLists.txt
+++ b/examples/CMakeLists.txt
@@ -6,45 +6,52 @@ set(CMAKE_CXX_STANDARD 11)
 add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/../crates/c-api ${CMAKE_CURRENT_BINARY_DIR}/wasmtime)
 
 function(CREATE_TARGET TARGET TARGET_PATH)
-	add_executable(wasmtime-${TARGET} ${TARGET_PATH})
+  add_executable(wasmtime-${TARGET} ${TARGET_PATH})
 
-	if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-		target_compile_options(wasmtime-${TARGET} PRIVATE -Wall -Wextra -Wno-deprecated-declarations)
-	elseif(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
-		target_compile_options(wasmtime-${TARGET} PRIVATE /W3)
-	endif()
+  if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    target_compile_options(wasmtime-${TARGET} PRIVATE -Wall -Wextra -Wno-deprecated-declarations)
+  elseif(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+    target_compile_options(wasmtime-${TARGET} PRIVATE /W3)
+  endif()
 
-        target_compile_definitions(wasmtime-${TARGET} PRIVATE WASMTIME_TEST_ONLY)
+  target_compile_definitions(wasmtime-${TARGET} PRIVATE WASMTIME_TEST_ONLY)
 
-	set_target_properties(wasmtime-${TARGET} PROPERTIES
-		OUTPUT_NAME wasmtime-${TARGET}
-		RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/$<0:>
-		CXX_VISIBILITY_PRESET hidden
-		POSITION_INDEPENDENT_CODE ON)
+  set_target_properties(wasmtime-${TARGET} PROPERTIES
+    OUTPUT_NAME wasmtime-${TARGET}
+    RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/$<0:>
+    CXX_VISIBILITY_PRESET hidden
+    POSITION_INDEPENDENT_CODE ON)
 
-	target_include_directories(wasmtime-${TARGET} PUBLIC wasmtime)
-	target_link_libraries(wasmtime-${TARGET} PUBLIC wasmtime)
-	if(APPLE)
-		target_link_libraries(wasmtime-${TARGET} PRIVATE "-framework CoreFoundation")
-	endif()
-	add_test(NAME ${TARGET}-c COMMAND $<TARGET_FILE:wasmtime-${TARGET}> WORKING_DIRECTORY ../..)
+  target_include_directories(wasmtime-${TARGET} PUBLIC wasmtime)
+  target_link_libraries(wasmtime-${TARGET} PUBLIC wasmtime)
+  if(APPLE)
+    target_link_libraries(wasmtime-${TARGET} PRIVATE "-framework CoreFoundation")
+  endif()
+  add_test(NAME ${TARGET}-c COMMAND wasmtime-${TARGET} WORKING_DIRECTORY ../..)
 endfunction()
 
 function(CREATE_RUST_TEST EXAMPLE)
-	if(ARGC GREATER 1)
-		add_test(NAME ${EXAMPLE}-rust COMMAND cargo run --example ${EXAMPLE} --features ${ARGV1} WORKING_DIRECTORY ../..)
-	else()
-		add_test(NAME ${EXAMPLE}-rust COMMAND cargo run --example ${EXAMPLE} WORKING_DIRECTORY ../..)
-	endif()
+  if(ARGC GREATER 1)
+    add_test(NAME ${EXAMPLE}-rust COMMAND cargo run --example ${EXAMPLE} --features ${ARGV1} WORKING_DIRECTORY ../..)
+  else()
+    add_test(NAME ${EXAMPLE}-rust COMMAND cargo run --example ${EXAMPLE} WORKING_DIRECTORY ../..)
+  endif()
 endfunction()
 function(CREATE_RUST_WASM EXAMPLE TARGET)
-	execute_process(COMMAND cargo build -p example-${EXAMPLE}-wasm --target ${TARGET})
+  add_custom_target(${EXAMPLE}-wasm-${TARGET} ALL COMMAND cargo build -p example-${EXAMPLE}-wasm --target ${TARGET})
 endfunction()
 
 # Enable testing
 enable_testing()
 
-# Add all examples
+# Wasm files required by tests
+create_rust_wasm(fib-debug wasm32-unknown-unknown)
+create_rust_wasm(tokio wasm32-wasip1)
+create_rust_wasm(wasi wasm32-wasip1)
+create_rust_wasm(wasi wasm32-wasip2)
+create_rust_wasm(component wasm32-unknown-unknown)
+
+# C/C++ examples/tests
 create_target(anyref anyref.c)
 create_target(async async.cpp)
 create_target(externref externref.c)
@@ -61,13 +68,8 @@ create_target(serialize serialize.c)
 create_target(threads threads.c)
 create_target(wasip1 wasip1/main.c)
 
-# Add rust tests
+# Rust examples/tests
 create_rust_test(anyref)
-create_rust_wasm(fib-debug wasm32-unknown-unknown)
-create_rust_wasm(tokio wasm32-wasip1)
-create_rust_wasm(wasi wasm32-wasip1)
-create_rust_wasm(wasi wasm32-wasip2)
-create_rust_wasm(component wasm32-unknown-unknown)
 create_rust_test(epochs)
 create_rust_test(externref)
 create_rust_test(fib-debug)

